src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,393 @@
+#include "split_and_merge_algorithm.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <random>
+
+#include "src/utils/distributions.h"
+
+void SplitAndMergeAlgorithm::read_params_from_proto(
+    const bayesmix::AlgorithmParams &params) {
+  BaseAlgorithm::read_params_from_proto(params);
+  n_restr_gs_updates = params.splitmerge_n_restr_gs_updates();
+  n_mh_updates = params.splitmerge_n_mh_updates();
+  n_full_gs_updates = params.splitmerge_n_full_gs_updates();
+}
+
+void SplitAndMergeAlgorithm::print_startup_message() const {
+  std::string msg = "Running Split and Merge algorithm with " +
+                    bayesmix::HierarchyId_Name(unique_values[0]->get_id()) +
+                    " hierarchies, " +
+                    bayesmix::MixingId_Name(mixing->get_id()) + " mixing...";
+  std::cout << msg << std::endl;
+}
+
+void SplitAndMergeAlgorithm::sample_allocations() {
+  auto &rng = bayesmix::Rng::Instance().get();
+
+  // MH updates
+  for (unsigned int k = 0; k < n_mh_updates; ++k) {
+    unsigned int n_data = data.rows();
+
+    // Sample first_random_idx and second_random_idx from the datapoints
+    Eigen::VectorXd probas = Eigen::VectorXd::Constant(n_data, 1.0 / n_data);
+    unsigned int first_random_idx = bayesmix::categorical_rng(probas, rng, 0);
+    probas = Eigen::VectorXd::Constant(n_data - 1, 1.0 / (n_data - 1));
+    unsigned int second_random_idx = bayesmix::categorical_rng(probas, rng, 0);
+    if (second_random_idx >= first_random_idx) {
+      second_random_idx++;
+    }
```

Comment:
What's the purpose of this block?

Reply:
From line 34 to 40 we have to sample two datapoints' indexes.
For the first index, we sample from 0 to n_data-1.
The second index will be an index from 0 to n_data-1, with the exception of the first index. Therefore, we sample from 0 to (n_data-1)-1, then, if the sampled value is greater than the first index, we increase it by one. In this way, it is like sampling from 0 to n_data-1, leaving a "hole" where before there was the first index.